waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -0,0 +1,606 @@
+        port0 pod-base-port
+        template-env (into [;; We set these two "MESOS_*" variables to improve interoperability
+                            {:name "MESOS_DIRECTORY" :value home-path}
+                            {:name "MESOS_SANDBOX" :value home-path}]
+                           (concat
+                             (for [[k v] common-env]
+                               {:name k :value v})
+                             (for [i (range ports)]
+                               {:name (str "PORT" i) :value (str (+ port0 i))})))
+        params (into {:k8s-name (service-id->k8s-app-name scheduler service-id)
+                      :backend-protocol backend-proto
+                      :backend-protocol-caps (string/upper-case backend-proto)
+                      :cmd cmd
+                      :cpus cpus
+                      :env template-env
+                      :grace-period-secs grace-period-secs
+                      :health-check-interval-secs health-check-interval-secs
+                      :health-check-max-consecutive-failures health-check-max-consecutive-failures
+                      :health-check-url (sd/service-description->health-check-url service-description)
+                      :home-path home-path
+                      :memory  (str mem "Mi")
+                      :min-instances min-instances
+                      :orchestrator-name orchestrator-name
+                      :port-count ports
+                      :replicaset-api-version replicaset-api-version
+                      :run-as-user run-as-user
+                      :service-id service-id
+                      :ssl? (= "https" backend-proto)}
+                     (for [i (range ports)]
+                       [(keyword (str "port" i)) (+ port0 i)]))
+        edn-opts {:readers {'waiter/param params
+                            'waiter/param-str (comp str params)}}]
+    (try
+      (->> replicaset-spec-file-path slurp (edn/read-string edn-opts))
+      (catch Throwable e
+        (log/error e "Error creating ReplicaSet specification for" service-id)
+        (throw e)))))
+
+(defn- create-service
+  "Reify a Waiter Service as a Kubernetes ReplicaSet."
+  [{:keys [service-id] :as descriptor}
+   {:keys [api-server-url http-client replicaset-api-version] :as scheduler}
+   service-id->password-fn]
+  (let [{:strs [run-as-user] :as service-description} (:service-description descriptor)
+        spec-json (service-spec scheduler service-id service-description service-id->password-fn)
+        request-url (str api-server-url "/apis/" replicaset-api-version "/namespaces/"
+                         (service-description->namespace service-description) "/replicasets")
+        response-json (api-request http-client request-url
+                                   :body (json/write-str spec-json)
+                                   :request-method :post)]
+    (replicaset->Service response-json)))
+
+(defn- delete-service
```

Comment:
How are we cleaning things up if this fails after we marked the app as `killed`?  Do we ignore `killed` apps somewhere?

Reply:
I changed the implementation to leverage K8s GC. The `delete-service` implementation is now much simpler, and the `waiter-app-status` attribute has been removed.